File: 2024/1/Cpp/solution.cpp
#include <iostream>
#include <vector> // std::vector<int> - data structure
#include <algorithm> // std:sort()
#include <cmath> // std::abs()

class Solution {
  private:
    std::vector<int> firstList;
    std::vector<int> secondList;

  int calcDistance(){
    int sum = 0;

    for (size_t i=0; i < firstList.size(); i++){
      sum += std::abs(firstList[i]-secondList[i]);
    }

    return sum;
  }

  public:

  Solution(std::vector<int> flist, std::vector<int> slist) {
      firstList = flist;
      secondList = slist;

      sort(firstList.begin(), firstList.end());
      sort(secondList.begin(), secondList.end());
  }

  ~Solution() {
  }

  void displayDistance(){
      int distance = calcDistance();
      std::cout << distance << std::endl;
  }
};



int main() {
    std::vector<int> vecFirstList = {3, 4, 2, 1, 3, 3};
    std::vector<int> vecSecondList = {4, 3, 5, 3, 9, 3};

    Solution solution(vecFirstList, vecSecondList);

    solution.displayDistance();

    return 0;
}

File: 2024/1/Python/solution.py
import numpy as np
from typing import List

FIRST_LIST = [3,4,2,1,3,3]
SECOND_LIST = [4,3,5,3,9,3]


class Solution():

  def __init__(self, flist, slist):
    if len(flist) == len(slist):
      flist.sort()
      slist.sort()
      self._first_list: List = flist
      self._second_list: List = slist
    else:
      raise Exception("Lists have to be the same length!")

  def get_lists_distance(self):
    return sum([abs(first_id-sec_id) for first_id, sec_id in zip(self._first_list, self._second_list)])
  

def numpy_get_lists_distance(flist, slist):   
  return sum(np.abs(np.sort(np.array(flist))-np.sort(np.array(slist))))


if __name__ == '__main__':
  '''Solution based on a numpy library'''
  print(numpy_get_lists_distance(FIRST_LIST, SECOND_LIST))

  '''Solution based on a python list manipulation'''
  solution = Solution(FIRST_LIST, SECOND_LIST)
  print(solution.get_lists_distance())